src/time/Waveform.cpp

```diff
@@ -20,110 +22,159 @@ int Waveform::getInterpolationOrder() const
   return _interpolationOrder;
 }
 
-void Waveform::initialize(
-    const Eigen::VectorXd &values)
+void Waveform::initialize(const Eigen::VectorXd &values)
 {
   int storageSize;
+  _timeStepsStorage = std::map<double, Eigen::VectorXd>{};
   PRECICE_ASSERT(_interpolationOrder >= Time::MIN_INTERPOLATION_ORDER);
-  storageSize            = _interpolationOrder + 1;
-  _timeWindowsStorage    = Eigen::MatrixXd::Zero(values.size(), storageSize);
-  _numberOfStoredSamples = 1; // the first sample is automatically initialized as zero and stored.
-  _storageIsInitialized  = true;
-  PRECICE_ASSERT(this->maxNumberOfStoredSamples() == storageSize);
-  PRECICE_ASSERT(this->valuesSize() == values.size());
-  for (int sampleIndex = 0; sampleIndex < maxNumberOfStoredSamples(); ++sampleIndex) {
-    this->storeAt(values, sampleIndex);
+  _timeStepsStorage[0.0] = Eigen::VectorXd(values);
+  _timeStepsStorage[1.0] = Eigen::VectorXd(values);
+}
+
+void Waveform::store(const Eigen::VectorXd &values, double normalizedDt)
+{
+  PRECICE_ASSERT(_timeStepsStorage.size() > 0);
+  // dt has to be in interval (0.0, 1.0]
+  PRECICE_ASSERT(normalizedDt > 0.0); // cannot override value at beginning of window. It is locked!
```

Comment:
Be aware of floating point values. Better use functions from `math::differences.hpp`